src/common/Fixed.ml

```diff
@@ -0,0 +1,236 @@
+  function `f` yielding a tuple of meta-data and pattern with elements of 
+  type `'r` to which `f` is further applied. 
+  
+  The `unfold` terminates when a pattern element which does not carry an `'r` is 
+  reached.
+
+  As with `fold_pattern` the function `f` returns a tuple of meta-data and 
+  pattern rather than our record type since, in general, `'r =/= 'a t`. *)
+
+  val rewrite_top_down : f:('a t -> 'a t) -> 'a t -> 'a t
+  (** `rewrite_top_down` specializes `unfold` by requiring that `'r = 'a t`. 
+  As a consequence the function `f` accepts our record type `'a t` as its 
+  argument. *)
+end
+
+(** Functor  which creates the fixed-point of the type defined in the `Pattern`
+module argument 
+*)
+module Make (Pattern : Pattern.S) : S with module Pattern := Pattern = struct
+  type 'a t = {pattern: 'a t Pattern.t; meta: 'a}
+  [@@deriving compare, map, fold, hash, sexp]
+
+  let fix (meta, pattern) = {meta; pattern}
+  let unfix {meta; pattern} = (meta, pattern)
+  let pattern_of {pattern; _} = pattern
+  let meta_of {meta; _} = meta
+
+  let rec pp f ppf {pattern; meta} =
+    Fmt.pf ppf {|%a%a|} f meta (Pattern.pp (pp f)) pattern
+
+  include Foldable.Make (struct type nonrec 'a t = 'a t
+
+                                let fold = fold
+  end)
+
+  let rec fold_pattern ~f t =
+    let meta, pattern = unfix t in
+    let pattern' = Pattern.map (fold_pattern ~f) pattern in
+    f (meta, pattern')
+
+  (** For clarity this is written explicitly but is equivalent to 
+  `fold_pattern ~f:(Fn.compose f fix) t` 
+  *)
+  let rec rewrite_bottom_up ~f t =
+    let x = {t with pattern= Pattern.map (rewrite_bottom_up ~f) t.pattern} in
+    f x
+
+  let rec unfold_pattern ~f x =
+    let meta, pattern = f x in
+    let pattern' = Pattern.map (unfold_pattern ~f) pattern in
+    fix (meta, pattern')
+
+  (** For clarity this is written explicitly but it is equivalent to
+  `unfold ~f:(Fn.compose unfix f) x` 
+  *)
+  let rec rewrite_top_down ~f x =
+    let t = f x in
+    {t with pattern= Pattern.map (rewrite_top_down ~f) t.pattern}
+end
+
+(** Nested fixed-point type where an element of the `Pattern` is itself
+a fixed-point type. We use this to represent statements which contain 
+expressions.
+*)
+module type S2 = sig
+  module First : S
+  module Pattern : Pattern.S2
+
+  type ('a, 'b) t = {pattern: ('a First.t, ('a, 'b) t) Pattern.t; meta: 'b}
+  [@@deriving compare, map, fold, hash, sexp]
+
+  val fix : 'b * ('a First.t, ('a, 'b) t) Pattern.t -> ('a, 'b) t
+  val unfix : ('a, 'b) t -> 'b * ('a First.t, ('a, 'b) t) Pattern.t
+  val pattern_of : ('a, 'b) t -> ('a First.t, ('a, 'b) t) Pattern.t
+  val meta_of : ('a, 'b) t -> 'b
+
+  include Foldable.S2 with type ('a, 'b) t := ('a, 'b) t
+  include Pretty.S2 with type ('a, 'b) t := ('a, 'b) t
+
+  val fold_pattern :
+       f:('a * 'r1 First.Pattern.t -> 'r1)
+    -> g:('b * ('r1, 'r2) Pattern.t -> 'r2)
+    -> ('a, 'b) t
+    -> 'r2
+  (** 
```

Comment:
Are all of these docstrings in the right place? Does this work with ocamldoc?

Reply:
I haven't used ocamldoc before so I don't know I'm afraid. I'll take a look at the docs for it. Are there examples elsewhere in the code base of how it should be done?

Reply:
I've just checked and I think they conform to ocamldoc conventions. I think ocamlformat enforces this (it certainly issues a warning about ambiguous associations of comments).